dash-pipeline/SAI/src/p4meta.h

```diff
@@ -0,0 +1,256 @@
+                _In_ const std::string &key_name) const;
+
+        const P4MetaKey* get_meta_object_key() const;
+
+        const P4MetaActionParam* get_meta_action_param(
+                _In_ uint32_t action_id,
+                _In_ sai_attr_id_t attr_id) const;
+
+        pi_p4_id_t find_action_id(
+                _In_ uint32_t attr_count,
+                _In_ const sai_attribute_t *attr_list) const;
+
+        uint32_t find_action_enum_id(
+                _In_ uint32_t action_id) const;
+    };
+
+
+    template<typename T>
+    static inline
+    void  set_attr_value_to_p4(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const sai_attribute_value_t &value,
+            _Inout_ T *p4_key_or_param)
+    {
+        if (field == "booldata")
+            return booldataSetVal(value, p4_key_or_param, bitwidth);
+        if (field == "u8")
+            return u8SetVal(value, p4_key_or_param, bitwidth);
+        if (field == "u16")
+            return u16SetVal(value, p4_key_or_param, bitwidth);
+        if (field == "s32")
+            return s32SetVal(value, p4_key_or_param, bitwidth);
+        if (field == "u32")
+            return u32SetVal(value, p4_key_or_param, bitwidth);
+        if (field == "u64")
+            return u64SetVal(value, p4_key_or_param, bitwidth);
+        if (field == "ipaddr")
+            return ipaddrSetVal(value, p4_key_or_param, bitwidth);
+        if (field == "mac")
+            return macSetVal(value, p4_key_or_param, bitwidth);
+        if (field == "u8list")
+            return u8listSetVal(value, p4_key_or_param, bitwidth);
+
+        assert(0);
+    }
+
+    void  set_attr_value_to_p4(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const sai_attribute_value_t &value,
+            _Inout_ p4::v1::FieldMatch_LPM *mf_lpm);
+
+    template<typename T>
+    static inline
+    void get_attr_value_from_p4(
+            _In_ const std::string &field,
+            _In_ uint32_t bitwidth,
+            _In_ const T *p4_key_or_param,
+            _Out_ sai_attribute_value_t &value)
+    {
+        const char *v = p4_key_or_param->value().c_str();
+
+        if (field == "booldata") {
+           value.booldata = *(const bool*)v;
+        }
+        else if (field == "u8") {
+           value.u8 = *(const uint8_t*)v;
+        }
+        else if (field == "u16") {
+           uint16_t val = *(const uint16_t*)v;
+           value.u16 = ntohs(val);
+        }
+        else if (field == "s32") {
+           int32_t val = *(const int32_t*)v;
+           value.s32 = ntohl(val) >> (32 - bitwidth);
+        }
+        else if (field == "u32") {
+           uint32_t val = *(const uint32_t*)v;
+           value.u32 = ntohl(val) >> (32 - bitwidth);
+        }
+        else if (field == "u64") {
+           uint64_t val = *(const uint64_t*)v;
+           if (*reinterpret_cast<const char*>("\0\x01") == 0) { // Little Endian
```

Comment:
what is that ?

Reply:
A quick approach to checking host arch endianness, from https://github.com/sonic-net/DASH/blob/main/dash-pipeline/SAI/src/utils.h#L143.

Reply:
please don't use that, use proper macros/functions defined endian.h? and make proper function for that if you need running time host endian